### PR TITLE
Iterate over each artifacts instead of capturing in a list

### PR DESCRIPTION
**Pull request by harshad16**
Iterate over each artifacts instead of capturing in a list
Signed-off-by: Harshad Reddy Nalla <hnalla@redhat.com>

## Related Issues and Dependencies

Related-to: https://github.com/thoth-station/thoth-application/issues/2690#issuecomment-1331752960

### Description

- Moved the function code `get_package_artifacts` directly into `get_package_hashes`
As each artifact object, is linked to the downloaded artifacts, they get deleted on obj deletion.
With this , each artifact would be _live_ for short period till the hash is gathered.

Each artifact would be iterated one after another, instead of gathering them all in one list.

- The function `get_package_data` didn't seem to be used anywhere
so it is now merged in `get_package_hashes`

### CODE REVIEW

**Review by mayaCostantini — APPROVED** [review]
lgtm, thanks! :100:

File: thoth/python/source.py [hunk 1]
```diff
@@ -414,38 +388,42 @@ def provides_package_version(self, package_name: str, package_version: str) -> b
...
 
     @lru_cache(maxsize=10)
-    def get_package_data(self, package_name: str, package_version: str, with_included_files: bool = False) -> list:
-        """Get information about release hashes and symbols available in this source index."""
+    def get_package_hashes(
+        self, package_name: str, package_version: str, with_included_files: bool = False, with_symbols: bool = False
+    ) -> list:
+        """Get information about release hashes available in this source index."""
         if self.warehouse:
             return self._warehouse_get_package_hashes(package_name, package_version, with_included_files)
```
Comment (KPostOffice) [inline thoth/python/source.py:396]: `with_symbols` is lost here

File: thoth/python/source.py [hunk 2]
```diff
@@ -414,38 +388,42 @@ def provides_package_version(self, package_name: str, package_version: str) -> b
             return False
 
     @lru_cache(maxsize=10)
-    def get_package_data(self, package_name: str, package_version: str, with_included_files: bool = False) -> list:
-        """Get information about release hashes and symbols available in this source index."""
+    def get_package_hashes(
```
Comment (KPostOffice) [inline thoth/python/source.py:391]: Why the change of the function name/doc line?

**Review by KPostOffice — COMMENTED** [review]
Just a couple of comments, otherwise I think this is a really good optimization.

File: thoth/python/source.py (hunk 2 above)
Comment (harshad16) [reply thoth/python/source.py:391]: I checked for the function `get_package_data` across whole github, it didn't seem be used anywhere.
so merged it in function `get_package_hashes`

File: thoth/python/source.py (hunk 1 above)
Comment (harshad16) [reply thoth/python/source.py:396]: it's not used by `_warehouse_get_package_hashes` right? did i miss something?

the only difference I noticed between `get_package_data` func and `get_package_hashes` func was the part 
`doc["symbols"] = artifact.get_versioned_symbols()`

thats the reason to include the param `with_symbols`
please correct me if i missed it somewhere

File: thoth/python/source.py (hunk 2 above)
Comment (KPostOffice) [reply thoth/python/source.py:391]: Ahh got it

File: thoth/python/source.py (hunk 1 above)
Comment (KPostOffice) [reply thoth/python/source.py:396]: Just took a look at the code. My worry was that the symbols were never added to the document if the flag was  set. It seems the opposite is actually true. `_warehouse_get_package_hashes` always adds the symbols not matter what. I think we should add a conditional statement in `_warehouse_get_package_hashes` so the return format is consistent. Let me know what you think

Comment (harshad16) [reply thoth/python/source.py]: ack, i have include the flag `with_symbols` in `_warehouse_get_package_hashes` as well.
Make sense to have consistency.

PTAL and Review again.
Thanks

**Review by KPostOffice — COMMENTED** [review]
/lgtm

**Comment by harshad16** [comment]
Merging it for release.
